### PR TITLE
don't let the the user disable encryption once it was activated

### PR DESCRIPTION
**Pull request by schiessle**
part of https://github.com/owncloud/core/issues/16021

We don't want to allow the user to disable encryption again after encryption was enabled, see also this discussion https://github.com/owncloud/core/pull/15880#discussion_r29142909

This is how it looks at the beginning, nothing changed:

After the admin enabled encryption he will see:

At this point the admin can decide to disable encryption again if he checked the checkbox by accident. If he decide to keep it and leave the page and come back later the setting will look like this

Now the checkbox is disabled and it is no longer possible to disable encryption

cc @DeepDiver1975 @MTRichards @jancborchardt

### CODE REVIEW

**Comment by jancborchardt** [comment]
@schiesbn the interaction is a tad bit strange, especially with the page reload as a limit for confirming. I would rather do the »Enable server-side encryption« as a button.

Then for confirmation, we have several options:
* Show the message in red as above and in addition add a button saying »Yes, enable encryption« which is the confirmation.
* Similar to how Github handles repo deletions, show the warning as well as »Please type "encryption" to confirm you really want to encrypt all data«, have them type it in an input box and press the »Encrypt« button next to it.

**Comment by jancborchardt** [comment]
(Checkboxes have no place in a one-way process.)

**Comment by jancborchardt** [comment]
Also cc @owncloud/designers ^

File: settings/css/settings.css [hunk 1]
```diff
@@ -494,3 +494,7 @@ doesnotexist:-o-prefocus, .strengthify-wrapper {
 #encryptionModules {
 	padding: 10px;
 }
+
+#warning {
+	color: red;
+}
```
Comment (MorrisJobke) [inline settings/css/settings.css]: new line please ;)

**Comment by schiesbn** [comment]
> (Checkboxes have no place in a one-way process.)

I had a button first. The bad thing about it is that after the encryption was enabled you would have basically the third screenshort without the line with the checkbox which could be kind of confusing whether encryption is enabled or not. That's why we thought a checkbox might be better, also as a indicator that encryption is enabled.

File: settings/css/settings.css (hunk 1 above)
Comment (schiesbn) [inline settings/css/settings.css]: It's already in the pipe, I just wanted to wait for Jenkins :smile:

**Comment by nickvergessen** [comment]
> Checkboxes have no place in a one-way process.

Just to underline what schiesbn already said. I think it makes sense to have a visual option to see that something was an option and was opt-ed in. It makes it way more visible, that someone decided to take the option, then having to scan for a little sentence on this page

**Comment by jancborchardt** [comment]
Ok, then a checkbox for the first step instead of a button. But we need some kind of confirmation as stated above. @MorrisJobke do you think the button or the input is better?

**Comment by jancborchardt** [comment]
And then as soon as the confirmation is done, the checkbox should change to being checked _and_ disabled.

**Comment by schiesbn** [comment]
> Ok, then a checkbox for the first step instead of a button. But we need some kind of confirmation as stated above.

Why a explicit confirmation? Isn't it enough to let the admin change his mind as long as he is on the settings page with a warning like above that this is your last chance to change your mind?

If you really insist hon having some kind of confirmation I would go for a button. Having a input field where people need to type something is way to much in my view. I don't think you can compare it with irrevocable deleting a github repository.

**Comment by MorrisJobke** [comment]
> I don't think you can't compare it with irrevocable deleting a github repository.

But it is irrecoverable, isn't it?



> And then as soon as the confirmation is done, the checkbox should change to being checked and disabled.

I like this. And only show the additional options once this is confirmed. Otherwise its weird, why you could disable the checkbox when you're still on the page, but not after the reload. What is the requirement for the disabled checkbox? Is the request send on page close? Or is it just because then this couldn't be done that far ago? keep in mind: you can leave the page open for hours ... what happens then?

**Comment by jancborchardt** [comment]
Exactly what @MorrisJobke said – having a page refresh as »confirmation« is highly unstable, as we saw through our first flawed implementation of undo deletion. (Same comment was also made by @DeepDiver1975 just now when we talked about this)

**Comment by schiesbn** [comment]
> having a page refresh as »confirmation« is highly unstable, as we saw through our first flawed implementation of undo deletion. 

Just to make this clear. Page refresh isn't used as confirmation. As soon as you check the checkbox encryption is enabled (same as all other settings work in ownCloud) we just let the admin disable it again as long as the page stays open. That's different to the flawed implementation of undo delete.

The only think we want to allow is "oh, I didn't want this... uncheck it again", that's why we don't disable the checkbox directly after checking it. But the settings gets updated directly.

**Comment by jancborchardt** [comment]
@schiesbn ok … so when does encryption start then?

**Comment by MTRichards** [comment]
We have a bit of a problem here, because what we have done is now made it impossible to stop encrypting files stored on the server. I know this is temporary, because in 8.2 we will provide a mechanism for disabling encryption and then batch unencrypting every file on the server (and the same for enabling encryption and batch encrypting files). HOWEVER...

We need two buttons: 
- Enable server side encryption (so that previously encrypted files can be read and updated)
- Turn on / off active encryption of files (so that all new and updated files will be encrypted)

Maybe they should just be:
- Encrypt all new or updated files
- Enable access to previously encrypted files (recommended)

The point is, it should be possible to turn off encryption - as it is implemented here this is not possible. That should not be the case - we should be able to turn off active file encryption, but keep the app enabled so that files that were encrypted can be decrypted after the app is turned off.

@jancborchardt @karlitschek

**Comment by schiesbn** [comment]
>    Enable server side encryption (so that previously encrypted files can be read and updated)

This is always possible. As long as the encryption module is enabled we will still detect encrypted files and read it correctly no matter if encryption is enabled or disabled.

>   Turn on / off active encryption of files (so that all new and updated files will be encrypted)

That's the "enable server-side encryption" checkbox. If it is checked encryption is active and works in both directions, read and write.

**Comment by schiesbn** [comment]
> ok … so when does encryption start then?

As soon as "server-side encryption is activated and at least one encryption module is loaded

**Comment by schiesbn** [comment]
@MTRichards 

> The point is, it should be possible to turn off encryption - as it is implemented here this is not possible. That should not be the case - we should be able to turn off active file encryption, but keep the app enabled so that files that were encrypted can be decrypted after the app is turned off.

This creates a lot of problem. Imagine you enabled encryption, encrypt some files and share some of them, now you disable encryption. You are still able to read the encrypted files as long as the encryption module is loaded. If you edit the file the encrypted file will be replaced with a unencrypted file because we no longer actively encrypt files. So far so good. Problems start if you rename/move a encrypted file or if you share a file which is still encrypted. Because the encryption keys will no longer move together with the file and if a file gets shared we will no longer re-encrypt the file with the new set of public keys.

That's why we decided for now the best option is that if you enable encryption it stays enabled. In oc8.2, when we have a occ command to decrypt all files again (under some circumstances) we can allow to disable encryption again. Because then the admin can disable it, decrypt all files and continue with ownCloud in a persistent state.

**Comment by MTRichards** [comment]
>Because the encryption keys will no longer move together with the file and if a file gets shared we will no longer re-encrypt the file with the new set of public keys.
Did not know this one. This will "lose" data. That is indeed a big problem. Is this also true for the recovery key - in other words, enable it at the beginning or not at all, or can that be changed over time?

The point then is that the customer MUST be informed that in 8.1, turning on encryption is a "from now until the end of time" step. We need to add a little explanation of this on the admin page, something like:

"Note: encryption is a one way process. Once encryption is enabled, all files from that point forward will be encrypted on the server and it will not be possible to disable encryption at a later date. Only enable encryption if you want all files to be encrypted permanently."

Completely agree with how this works and the plan to fix this in 8.2:
>That's why we decided for now the best option is that if you enable encryption it stays enabled. In oc8.2, when we have a occ command to decrypt all files again (under some circumstances) we can allow to disable encryption again. Because then the admin can disable it, decrypt all files and continue with ownCloud in a persistent state.

That would be the follow up.

@carlaschroder we really need to document this, because encryption in 8.1 is now a 1 way process and cannot be turned off. We will enable it to be turned off in 8.2 with proper migration scripts.

**Comment by schiesbn** [comment]
> The point then is that the customer MUST be informed that in 8.1, turning on encryption is a "from now until the end of time" step. We need to add a little explanation of this on the admin page, something like

Not necessarily until the end of time but at least until they update to 8.2 in roughly 3 months :wink: 
But I agree that the user needs to be informed, that's the whole purpose of this pull request. You can see this on the second screenshot if you click on it so that you can read the text.

If the admin enabled encryption he gets a "warning": "Once encryption is enabled there is no way to disable it again. This is your last chance to disable it again.". Now the admin can change the settings by unchecking server-side encryption again. If he still think encryption is what he wants he can just move on, encryption will be enabled from now on.

**Comment by schiesbn** [comment]
@MTRichards sure, I can adjust the text

**Comment by MorrisJobke** [comment]
> > ok … so when does encryption start then?
> 
> As soon as "server-side encryption" is enabled and at least one encryption module is loaded

And if the admin keeps the page open for 5 minutes, in this time a user logs in and encrypts stuff and then the admin decides to disable encryption: what happens? (just having sync clients in mind - instance with 100 users - each client checks once in 30s - it isn't that unlikely)

I like it more, that you need a second confirmation to enable encryption and not being in a inbetween state just for the sake of undo.

**Comment by schiesbn** [comment]
> And if the admin keeps the page open for 5 minutes, in this time a user logs in and encrypts stuff and then the admin decides to disable encryption: what happens?

I consider this more a theoretical problem. Also I find it highly unintuitive if you have a state where "server side encryption" is checked (which means enabled in all other settings) but a button below where you have to click "yes" a second time. But I will add it if all of you think that this is how it should work.

**Comment by schiesbn** [comment]
Button added and the text was adjusted like requested by @MTRichards (with slight changes due to the new button). This is the text the admin will see now: 

```
Encryption is a one way process. Once encryption is enabled, all files from that point forward will be encrypted on the server and it will not be possible to disable encryption at a later date. This is the final warning: Do you really want to enable encryption? [Enable encryption]
```

( [Enable encrption] is the new button)

**Comment by MorrisJobke** [comment]
> I consider this more a theoretical problem.

No ...just have an instance with only 300 clients connected - that means 10 requests per second. Even if the admin immediately uncheck the checkbox - it will have an impact for the user. Please do it properly otherwise we will see weird stuff in bigger installations.

**Comment by MorrisJobke** [comment]
Thx @schiesbn

**Comment by MorrisJobke** [comment]
I fixed the label ID (now you can also click on the label and the checkbox switches the state - was a regression in this branch) and I added a proper style for the labels of disabled inputs. Now they are grey and doesn't switch the color on hover/focus. cc @jancborchardt 

This PR is a :+1: from my side. I tested it.

**Comment by MorrisJobke** [comment]
@schiesbn Can you brief @carlaschroder about the changes regarding the 8.1 encryption documentation. Thanks :)

**Comment by schiesbn** [comment]
@jancborchardt can you also have a final look at it so that we can merge it? Thanks!

**Comment by jancborchardt** [comment]
Looks good now :+1:

**Pull request merged**